### ISSUE TITLE
Extend `names_and_types` to contain type hashes

### ISSUE THREAD

**Issue by emersonknapp** [ros2/rmw#356]
#### Feature description

REP-2016 type hashes now exist and are passed during discovery for the 3 DDS implementations, but are only exposed through `rmw_topic_endpoint_info_t`, which is given by `get_publishers_info_by_topic`/`get_subscriptions_info_by_topic`, but otherwise is not exposed.

When querying information about a _node_ rather than a _topic_, one generally uses `rmw_get_publisher_names_and_types_by_node`(etc) which returns the topic name and typename, but does have access to the type hash, which it would be appropriate to return in these functions, making it easier to then query information about those topics.

#### Implementation considerations

This probably looks like:

```
typedef struct RMW_PUBLIC_TYPE rmw_names_and_types_s
{
  /// Array of names
  rcutils_string_array_t names;

  /// Dynamic array of arrays of type names, with the same length as `names`
  rcutils_string_array_t * types;

  /// NEW FIELD
  /// Dynamic array of type hashes, with the same length as `names`
  rosidl_type_hash_t * type_hashes;
} rmw_names_and_types_t;

```

And then will require filling out this information in all `names_and_types` APIs.

This would be ABI-breaking, so probably can only be planned for Jazzy.

**Comment by wjwwood** [comment]
Makes sense to me, basically anywhere we have the type of a topic we should have the hash.

**Comment by fujitatomoya** [comment]
sounds good to me.

**Comment by amparo-siemens** [comment]
I need the hash to obtain the information of an interface in runtime using the GetTypeDescription service. With the type hash discovery for msg is really easy since just executing the "ros2 topic info -v /topic" I obtain the required hash to call the service. But to use the GetTypeDescription with actions and services I need to check the hash manually in the install folder, so it is only possible if you have access to the source code. It could be really interesting to include this feature request to discover the hash for actions and services to be able to take the advantage of knowing the structure of the interfaces in runtime (without access to the code) using the GetTypeDescription service.

**Comment by gramss** [comment]
https://github.com/siemens/rosita would also benefit from this function.

If this is not picked up by other maintainers, we could try building a PR on our own. Although, guidance from existing maintainers would be highly welcome. The description by @emersonknapp could be already all needed information to start this. Still, happy to connect and get some pitfalls out of the way, before we start this effort on our site.

**Comment by wjwwood** [comment]
@gramss thanks for bumping this, we looked at it in the issue triage meeting and we agreed this makes sense, but that none of the maintainers have time to look at it so help is definitely appreciated. If you need help, feel free to join the Client Library working group (see the calendar of events, it's usually on Fridays: https://calendar.google.com/calendar/u/0/embed?src=agf3kajirket8khktupm9go748@group.calendar.google.com&ctz=America/Los_Angeles and the calendar event has a link to the video call and the agenda doc if you want to put yourself on there) or if failing that the ROS 2 PMC meeting if you can't make the other one with any questions you have. Or feel free to ask questions here via text.

**Comment by tisomi** [comment]
Hey everyone,

I was asked by @gramss to work on this issue. I’ve already started exploring the repos, and I’d like to clarify a few points before I go deeper:

1. The definition and initialization of the struct `rmw_names_and_types_t` live in the rmw repo, but the actual population of this struct happens in `rmw_dds_common` (`graph_cache.cpp`). Now, is it sufficient to extend the population logic in `graph_cache.cpp` (e.g. in `__get_names_and_types_from_gids` / `__populate_rmw_names_and_types`) to also handle `type_hashes`, or are there other places that also need to be updated?
2. Where exactly do I get the type_hashes of the different topics? Is there a function for that?
3. For testing, what’s the recommended way to validate that this works?

**Comment by tisomi** [comment]
Update

I’ve dug deeper into the functionality of the `rmw_dds_common` repo as well as the `rmw_fastrtps` repo.

So far, I’ve found that the functions `rmw_get_publisher_names_and_types_by_node`, `rmw_get_subscriber_names_and_types_by_node`, … are structured like the following graph:

<img width="2129" height="1059" alt="Image" src="https://github.com/user-attachments/assets/7edf45bf-53cd-4a21-9ac0-c643d8e77ccb" />

The `rmw_names_and_types_t`, which is the main container for the hashes, is populated deep down that path and then passed onward through all the functions and methods.

Question 2

> Where exactly do I get the type_hashes of the different topics? Is there a function for that?

It looks like `__get_names_and_types_from_gids` already receives the `EntityGidToInfo` struct, which contains the [`EntityInfo`](https://docs.ros.org/en/rolling/p/rmw_dds_common/generated/structrmw__dds__common_1_1EntityInfo.html#_CPPv4N14rmw_dds_common10EntityInfoE).
So it may already be possible to access the hashes without much extra overhead. But I still need to confirm this.

Question 3

> For testing, what’s the recommended way to validate that this works?

After discussing with @gramss, we think the safest approach is to build ROS from scratch in Docker and manually replace the affected repos. From there, I’d run some simple “hello world” nodes to check basic functionality.

In addition, I’ll write some gmock tests (though I’m new to gmock, so I’ll need to look into that first).

For testing strategy in particular, I’d really appreciate feedback from people with more experience — does this plan sound reasonable, or are there better alternatives?

**Comment by tisomi** [comment]
Hey guys, I've picked up this problem again and made some significant progress. On my fork, I can now use the `rmw_get_*_names_and_types_by_node` functions to access the type hashes. I've tested it successfully on Fast RTPS and Cyclone DDS. Zenoh and Connext DDS require more work.. I can't make a PR yet due to company restrictions.

**Comment by gramss** [comment]
short update @tisomi is working out the logistics of adding a PR from OSRA requirements and on top of that Siemens.
We love processes.. but are still optimistic to merge that at some point. Latest in summer, when @tisomi is working in a longer stretch with us.

**Comment by gramss** [comment]
hi @wjwwood,

From our point of view this looks good.
We would appriciate your time to review our 3 PRs for this new feature.

Thanks a lot!
